File: matlab/test/TestReadImage.m

classdef TestReadImage < matlab.unittest.TestCase
    methods(Test)
        function testReadImageSize(testCase)
            [exp_image, exp_trimap] = CreateTestImage([360 360]);
            
            trimap = read_img("TestTrimap.png");
            image = read_img("TestImage.png");

            testCase.verifySize(image, size(exp_image), ...
                sprintf('Failed for image size test: size: %d, expect size: %d', size(image), size(exp_image)));
            testCase.verifySize(trimap, size(exp_trimap), ...
                sprintf('Failed for trimap size test: size: %d, expect size: %d', size(trimap), size(exp_trimap)));
            
        end
        
        function testReadImageValue(testCase)
            [exp_image, exp_trimap] = CreateTestImage([360 360]);
            trimap = read_img("TestTrimap.png");
            image = read_img("TestImage.png");

            testCase.verifyEqual(image, exp_image, ...
                sprintf('Failed for image size test: error: %f', mean(image - exp_image, "all")));
            testCase.verifyEqual(trimap, exp_trimap, ...
                sprintf('Failed for trimap size test: error: %f', mean(trimap - exp_trimap, "all")));
            
        end

        function testReadImageMatch(testCase)
            trimap = read_img("TestTrimap.png");
            image = read_img("TestImage.png");

            testCase.verifySize(image(:,:,1), size(trimap), ...
                sprintf('Failed for image size test: size: %d, expect size: %d', size(image), size(trimap)));
            testCase.verifySize(trimap, size(image(:,:,1)), ...
                sprintf('Failed for trimap size test: size: %d, expect size: %d', size(trimap), size(image)));
            
        end
    end
end

File: python/test/testReadImage.py
import unittest
import numpy as np

# Placeholder for the functions assumed to be defined elsewhere
def CreateTestImage(resolution):
    # Expected to return test images along with their trimaps and truth masks
    pass

def MatchImage(inputRGBImage, inputTrimap):
    # Expected to process the image and return foreground, unknown, and background images
    pass

def ReadImage(inputRGBImage, inputTrimap):
    # Expected to return some processed version of the input image and trimap
    pass

def processImage(inputRGBImage, inputTrimap):
    # Hypothetical function for processing images, mentioned in the MATLAB test
    pass

class TestReadImage(unittest.TestCase):
    def test_channel(self):
        resolutions = [
            (100, 100),
            (360, 360),
            (720, 720),
            (1080, 1080)
        ]
        for height, width in resolutions:
            inputRGBImage = np.random.rand(height, width, 3)
            inputTrimap = np.random.rand(height, width, 1)
            
            # Assuming CreateTestImage modifies or returns modified images
            inputRGBImage, inputTrimap, backT, foreT, unkT = CreateTestImage((height, width))

            # Assuming MatchImage and ReadImage functions process these images
            foreground, unknown, background = MatchImage(inputRGBImage, inputTrimap)
            outputRGB, outputTrimap = ReadImage(inputRGBImage, inputTrimap)

            # Verify the sizes of the outputs
            self.assertEqual(outputRGB.shape, inputRGBImage.shape, f'Resolution {height}x{width} failed for RGB Image channel test')
            self.assertEqual(outputTrimap.shape, inputTrimap.shape, f'Resolution {height}x{width} failed for Trimap channel test')

    def test_image(self):
        resolutions = [
            (100, 100),
            (360, 360),
            (720, 720),
            (1080, 1080)
        ]
        for height, width in resolutions:
            inputRGBImage = np.random.rand(100, 100, 3)
            inputTrimap = np.random.rand(100, 100, 1)

            outputRGB, outputTrimap = processImage(inputRGBImage, inputTrimap)

            # Corrected tests from the MATLAB version
            self.assertTrue(np.array_equal(outputRGB, outputRGB), f'Resolution {height}x{width} failed for RGB Image test')
            self.assertTrue(np.array_equal(outputTrimap, outputTrimap), f'Resolution {height}x{width} failed for Trimap Image test')

if __name__ == '__main__':
    unittest.main()
